common/thorhelper/thorread.cpp

```diff
@@ -25,13 +25,152 @@
+                areMatchingPTrees(options, other->queryOptions());
+    }
+
+    virtual bool expectedMatchesProjected() const
+    {
+        return (expectedCrc && expectedCrc == projectedCrc) || (expectedMeta == projectedMeta);
+    }
+
+protected:
+    void ensureTranslators() const;
+
+protected:
+    RecordTranslationMode mode;
+    mutable bool checkedTranslators = false;
+    StringAttr format;
+    unsigned actualCrc;
+    unsigned expectedCrc;
+    unsigned projectedCrc;
+    Linked<IOutputMetaData> actualMeta;
+    Linked<IOutputMetaData> expectedMeta;
+    Linked<IOutputMetaData> projectedMeta;
+    Linked<const IPropertyTree> options;
+    mutable Owned<const IDynamicTransform> translator;
+    mutable Owned<const IKeyTranslator> keyedTranslator;
+};
+
+static CriticalSection translatorCS;
+void DiskReadMapping::ensureTranslators() const
+{
+    const char * filename = ""; // not known at this point
+    CriticalBlock block(translatorCS);
+    if (checkedTranslators)
+        return;
+
+    checkedTranslators = true;
+    IOutputMetaData * sourceMeta = expectedMeta;
+    unsigned sourceCrc = expectedCrc;
+    if (mode != RecordTranslationMode::AlwaysECL)
+    {
+        if (actualCrc && actualMeta)
+        {
+            sourceMeta = actualMeta;
+            sourceCrc = actualCrc;
+        }
+
+        if (actualCrc && expectedCrc && (actualCrc != expectedCrc) && (RecordTranslationMode::None == mode))
+            throwTranslationError(actualMeta->queryRecordAccessor(true), expectedMeta->queryRecordAccessor(true), filename);
+    }
+
+    //This has a very low possibility of Meta crcs accidentally matching, which could lead to a crashes on an untranslated files.
+    const RtlRecord & projectedRecord = projectedMeta->queryRecordAccessor(true);
+    const RtlRecord & sourceRecord = sourceMeta->queryRecordAccessor(true);
+    if (strsame(format, "csv"))
+    {
+        type_vals format = options->hasProp("ascii") ? type_string : type_utf8;
+        translator.setown(createRecordTranslatorViaCallback(projectedRecord, sourceRecord, format));
+    }
+    else if (strsame(format, "xml"))
+    {
+        translator.setown(createRecordTranslatorViaCallback(projectedRecord, sourceRecord, type_utf8));
+    }
+    else
+    {
+        if ((projectedMeta != sourceMeta) && (projectedCrc != sourceCrc))
+            translator.setown(createRecordTranslator(projectedRecord, sourceRecord));
+    }
+
+    if (translator)
+    {
+        if (!translator->canTranslate())
+            throw MakeStringException(0, "Untranslatable record layout mismatch detected for file %s", filename);
+
+        if (translator->needsTranslate())
+        {
+            if (sourceMeta != expectedMeta)
+            {
+                Owned<const IKeyTranslator> _keyedTranslator = createKeyTranslator(sourceMeta->queryRecordAccessor(true), expectedMeta->queryRecordAccessor(true));
+                //MORE: What happens if the key filters cannot be translated?
```

Comment:
I don't know, but probably should work out before finally enabling this code... Does it need a JIRA to revisit?